vertx-web/src/main/java/io/vertx/ext/web/templ/TemplateEngine.java

```diff
@@ -33,12 +33,17 @@
 public interface TemplateEngine {
 
   /**
-   * Render
+   * Render the template
+   * <p>
+   * <b>NOTE</b> if you call method directly (i.e. not using {@link io.vertx.ext.web.handler.TemplateHandler}) make sure
+   * that <i>templateFileName</i> is sanitized via {@link io.vertx.ext.web.impl.Utils#normalizePath(String)}
+   *
    * @param context  the routing context
-   * @param templateFileName  the template file name to use
+   * @param templateDirectory  the template directory to use
+   * @param templateFileName  the relative template file name to use
    * @param handler  the handler that will be called with a result containing the buffer or a failure.
    */
-  void render(RoutingContext context, String templateFileName, Handler<AsyncResult<Buffer>> handler);
+  void render(RoutingContext context, String templateDirectory, String templateFileName, Handler<AsyncResult<Buffer>> handler);
```

Comment:
Hm, I think this would work in most cases - I'm currently unable to test, can write tests in the evening (CET).

So we only look at users that call `render` directly and use _includes_. Presume we have template _index.hbs_ that includes _include.hbs_:

1. if its called with relative path to template `render(ctx, "templates/index", h)`, include is now done with `{{>templates/include}}`, after update this include is resolved to _templates/templates/include.hbs_ which breaks it, but this is actual problem (#514) we are trying to fix
2. if its called with absolute path `render(ctx, "/absolute/templates/index", h)`, include must also be done with full path `{{>/absolute/templates/include}}`, resolved path to _include.hbs_ will start with double slash `//absolute/..` but this does not break it
3. if its called with relative or absolute path on Windows `render(ctx, "C:\\absolute\\templates\\index", h)`, include is done as above, but since `/` is not used _templateDirectory_ is empty String and path to _include.hbs_ will start with / and that will break it ... but don't know what projects attitude for support on Windows system is?